projectorganizer/src/prjorg-goto-panel.c

```diff
@@ -0,0 +1,496 @@
+/*
+ * Copyright 2023 Jiri Techet <techet@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+/* This file contains mostly stolen code from the Colomban Wendling's Commander
+ * plugin. Thanks! */
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "prjorg-goto-panel.h"
+
+#include <gtk/gtk.h>
+#include <geanyplugin.h>
+
+
+enum {
+	COL_ICON,
+	COL_LABEL,
+	COL_PATH,
+	COL_LINENO,
+	COL_COUNT
+};
+
+
+//TODO: free on plugin unload
+struct {
+	GtkWidget *panel;
+	GtkWidget *entry;
+	GtkWidget *tree_view;
+	GtkListStore *store;
+} panel_data = {
+	NULL, NULL, NULL, NULL
+};
+
+
+static PrjorgGotoPanelLookupFunction lookup_function;
+
+extern GeanyData *geany_data;
+
+
+void prjorg_goto_symbol_free(PrjorgGotoSymbol *symbol)
+{
+	g_free(symbol->name);
+	g_free(symbol->file_name);
+	g_free(symbol->scope);
+	g_free(symbol->tooltip);
+	g_free(symbol);
+}
+
+
+static struct
+{
+	const gchar *icon_name;
+	GdkPixbuf *pixbuf;
+}
+/* keep in sync with Geany */
+geany_icons[_N_ICONS] = {
+	[_ICON_CLASS]		= { "classviewer-class", NULL },
+	[_ICON_MACRO]		= { "classviewer-macro", NULL },
+	[_ICON_MEMBER]		= { "classviewer-member", NULL },
+	[_ICON_METHOD]		= { "classviewer-method", NULL },
+	[_ICON_NAMESPACE]	= { "classviewer-namespace", NULL },
+	[_ICON_OTHER]		= { "classviewer-other", NULL },
+	[_ICON_STRUCT]		= { "classviewer-struct", NULL },
+	[_ICON_VAR]			= { "classviewer-var", NULL },
+};
+
+
+static GdkPixbuf *get_tag_icon(const gchar *icon_name)
+{
+	static GtkIconTheme *icon_theme = NULL;
+	static gint x = -1;
+
+	if (G_UNLIKELY(x < 0))
+	{
+		gint dummy;
+		icon_theme = gtk_icon_theme_get_default();
+		gtk_icon_size_lookup(GTK_ICON_SIZE_MENU, &x, &dummy);
+	}
+	return gtk_icon_theme_load_icon(icon_theme, icon_name, x, 0, NULL);
+}
+
+
+static GdkPixbuf *get_icon_pixbuf(gint icon)
+{
+	if (!geany_icons[_ICON_CLASS].pixbuf)
+	{
+		guint i;
+		for (i = 0; i < G_N_ELEMENTS(geany_icons); i++)
+			geany_icons[i].pixbuf = get_tag_icon(geany_icons[i].icon_name);
+	}
+
+	if (icon < _N_ICONS)
```

Comment:
I'd prefer if https://github.com/geany/geany/pull/3849/commits/1931fcc2f4d5de46871c85a782903d8c4d41e699 from https://github.com/geany/geany/pull/3849 could be merged and used here. While the icons can be accessed this way, if someone reorganizes something in Geany, it will break. So better have some proper API for accessing these icons.